cases: CI and ML [new]

drorata opened the pull request.

drorata commented on content/docs/use-cases/ci-and-ml.md:
```diff
@@ -0,0 +1,171 @@
+environments, or simply move the testing load from a developer’s local machine
+to a dedicated testing platform.
+
+![](/img/ci_for_ml_retrain_model.jpg) _Model training flow_
```
The image has to be redone (and centered)

drorata commented on content/docs/use-cases/ci-and-ml.md:
```diff
@@ -0,0 +1,171 @@
+
+      - name: Track the newly trained model
+        run: |
+          dvc add model.pkl
+          git config --global user.name 'Your Name'
+          git config --global user.email 'my@enail.com'
+          git add model.pkl.dvc
+          git commit -m "Model retrained upon sync to PR"
+          git push
+          dvc phsh
```
I am not sure whether the details here work as I expect them to. But I do hope the idea is clear.

jorgeorpinel replied:
Typo: `phsh` 🙂 

BTW what language is this? Github CI/CD? Maybe worth mentioning.

drorata replied:
Typo indeed! Will fix. Thanks.

That's a YMAL for GitHub Actions. I do mention at the beginning that the snippets are only for demonstration purposes and assume usage of GitHub. I based on https://cml.dev/.

Otherwise? @jorgeorpinel feel free to resolve this.

jorgeorpinel replied:
I see the Remarks now, thanks.

drorata commented on content/docs/use-cases/ci-and-ml.md:
```diff
@@ -0,0 +1,171 @@
+
+## Seamless model evaluation
+
+![](/img/ci_for_ml_evaluate_models.jpg) _Evaluating different models against the
+same validation dataset_
```
Redo image (and center)

drorata commented on content/docs/use-cases/ci-and-ml.md:
```diff
@@ -0,0 +1,171 @@
+          pip install -r requirements.txt
+
+      - name: DVC fetch the data used by the branch
+        run: dvc pull
```
Probably this line has to be refactored so it will fetch the right validation data set.

drorata replied:
@pared I think here I need a little help in and make it clearer that which version of the data is brought to the workspace.

pared replied:
To make it more readable I would consider `pull`ing specified target. To reproduce, I presume only data is needed (depending on how the project is structured). So, what we could do is `dvc pull data_eval.dvc` and then, in repro `dvc repro {eval_stage_name}` to stress where is actually important here.

jorgeorpinel commented on content/docs/sidebar.json:
```diff
@@ -75,7 +75,11 @@
+      "data-registries",
+      {
+        "label": "CI and ML",
+        "slug": "ci-and-ml"
```
Is is "and" or "for"?

drorata replied:
I had "and" in mind in the sense of storytelling; the story of how CI and ML play nicely together. It's a matter of taste, and "and" can be easily used and be more down to earth.

jorgeorpinel replied:
I mean the title in the actual doc uses "for". Up to you which one but let's just use the same in nav vs doc please.

drorata replied:
Sure. My bad...

jorgeorpinel commented on content/docs/use-cases/ci-and-ml.md:
```diff
@@ -0,0 +1,171 @@
+## Motivation
+
+ML engineers and data scientists face many unique challenges. Some of these
+challenges can be solved by adopting first-citizen concepts from traditional
```
What are "first-citizen concepts"? Not familiar with the phrase 🙂 is it similar to "first-class citizen"?

drorata replied:
😲 True! I'll replace it with: "... by adopting well-established concepts from...". @jorgeorpinel what do you think?

jorgeorpinel replied:
Sure, I honestly thought maybe it was a phrase I didn't know. If it's a typo feel free to replace it with something else like the above example. Consider term "best practices" maybe.

jorgeorpinel commented on content/docs/use-cases/ci-and-ml.md:
```diff
@@ -0,0 +1,171 @@
+challenges can be solved by adopting first-citizen concepts from traditional
+software engineering and development. Probably the most common adoption is
+versioning control. Some other, unfortunately less common, could be unit testing
+data processing steps of a pipeline. Here is a non exhaustive list of a few
```
"data processing steps of a pipeline" -> "defining reproducible workflows or pipelines" ?

drorata replied:
I do refer here to the need of implementing unit tests aimed at testing the logic in the data processing. I'm not sure I understand @jorgeorpinel point...

jorgeorpinel replied:
Oh, I see, I got it wrong. "unit testing data processing steps of a pipeline" is a little hard to read I guess.

jorgeorpinel replied:
Maybe just "data processing stages"? ("stage" is a more DVC-specific term, which you can even link to https://dvc.org/doc/command-reference/run if you want.) You can probably use the key work "pipeline" in other places

drorata replied:
I rephrased the sentence; let me know if it is easier to understand. I didn't link to `dvc run`, trying to keep it simpler.

jorgeorpinel commented on content/docs/use-cases/ci-and-ml.md:
```diff
@@ -0,0 +1,171 @@
+data processing steps of a pipeline. Here is a non exhaustive list of a few
+other unique challenges:
+
+- Automatically re-train models upon changes.
+- Compare (and report) performance of different versions of a model.
+- Test data correctness when new data is introduced.
+- Handle a model training that is computationally intensive.
```
Minor: these aren't complete sentences so no need for period `.`, I think.

drorata replied:
Based on [rule #1 here](https://writingcooperative.com/when-and-when-not-to-insert-punctuation-at-the-end-of-a-bullet-point-cefd3915950d) I would say that the punctuation is correct. But maybe I misunderstand it.

jorgeorpinel replied:
I can't open that because of Mediums article reading limits... Can you paste the rule? According to this one they shouldn't have periods AFAIU.

We do have guidelines around this in https://dvc.org/doc/user-guide/contributing/docs#doc-style-guidelines-javascript-and-markdown, please check out that doc if you haven't.

jorgeorpinel replied:
(this one = https://www.free-power-point-templates.com/articles/rules-for-using-periods-with-bullet-points/)

jorgeorpinel replied:
If it was a blog post I wouldn't care at all but we do have lots of other bullet lists in existing docs so the main thing here is to keep consistency with existing docs. Or if other docs are wrong, to open an issue or small PR to fix them 🙂

drorata replied:
Thanks for the pointer!

I really don't know. I feel comfortable with capitalization and punctuation as it is now. I'm also fine with changing it.

Here is the text from the [link](https://writingcooperative.com/when-and-when-not-to-insert-punctuation-at-the-end-of-a-bullet-point-cefd3915950d):

> Use a period at the end of a bullet point when the bullet point:
Completes a sentence based on its stem.
Is a complete sentence on its own.
The “stem” is made up of the words that end with a colon in the introduction to the bullet points. In the above statement, “Use a period at the end of a bullet point when the bullet point:” is the stem. Since each bullet point in the series that follows the stem completes the sentence begun by the stem, each ends with a period.

BTW and off-topic: in https://github.com/iterative/dvc.org/blame/master/content/docs/user-guide/contributing/docs.md#L185-L187. I think that `commit` is git oriented while `revision` and `reference` are maybe more SVN? I would say that the most neutral term is `version`.... But as I said. off-topic.

jorgeorpinel replied:
Thanks (I was able to open it with an incognito windows too).

These bullet items don't complete the sentence. It would need to be something like:

- this. `  # It would need to be something like this.` ✅ 
- or that. `  # It would need to be something like that.` ✅ 

But not when you say one of the following:

- one `  # But not when you say one of the following one?` ❌ 
- the other `  # But not when you say one of the following the other?` ❌ 

The difference is subtle and this is definitely too deep haha but yeah if you don't mind please don't use periods here.

jorgeorpinel replied:
> commit is git oriented while revision and reference are maybe more SVN? I would say that the most neutral term is version

Maybe all those terms are both in Git and other SCM tools. Commit being the more basic one but indeed "version" is even more general. We use commit extensively throughout docs since we expect 99% of our users know what that is. Maybe that's a wrong assumption, I'm not sure TBH. So far no complaints 🙂

jorgeorpinel commented on content/docs/use-cases/ci-and-ml.md:
```diff
@@ -0,0 +1,171 @@
+other unique challenges:
+
+- Automatically re-train models upon changes.
+- Compare (and report) performance of different versions of a model.
```
> Compare (and report) performance...

Maybe the challenge isn't that directly but rather doing it systematically or automatically?

drorata replied:
I'll add "systematically".

jorgeorpinel replied:
Cool

jorgeorpinel commented on content/docs/use-cases/ci-and-ml.md:
```diff
@@ -0,0 +1,171 @@
+
+- Automatically re-train models upon changes.
+- Compare (and report) performance of different versions of a model.
+- Test data correctness when new data is introduced.
```
> Test data ... data

Minor: sounds a bit redundant. Maybe use term "dataset" somehow in the 2nd instance? No biggie here, can be left as is (feel free to resolve this comment).

drorata replied:
Changed :)

jorgeorpinel replied:
Thanks!

jorgeorpinel commented on content/docs/use-cases/ci-and-ml.md:
```diff
@@ -0,0 +1,171 @@
+- Automatically re-train models upon changes.
+- Compare (and report) performance of different versions of a model.
+- Test data correctness when new data is introduced.
+- Handle a model training that is computationally intensive.
```
> Handle a model training...

What is the challenge? Not sure I understand this one 🙂

drorata replied:
This point is derived from a discussion I had with @shcheklein. It points at the case of a model training that either is very lengthy or depends on some specific environment (e.g. GPU) etc. In this case, the CI flow can run the training on the most suitable environment and free the developer's resources in the meantime. Does it make more sense?

drorata replied:
I take it that this one can be resolved.

drorata replied:
@jorgeorpinel Can resolve this one?

jorgeorpinel replied:
Yes. Sorry took so long, got distracted with a million things. Feel free to resolve the stuff you've addressed. Resolving this one. Thanks

jorgeorpinel commented:
Woa nice @drorata thanks for this! Left some first few comments but need to make a more comprehensive review soon.

jorgeorpinel commented:
p.s. the tests are failing but it's actually OK: our link checker gets confused when we add new docs or images and refer to them from other docs. Let's just double check the tests before merging this to confirm there's no other issue.

drorata commented:
@jorgeorpinel thanks for the initial feedback!

I don't know how to check the tests. On my local machine, I was able to run `yarn develop` and everything worked smoothly. I don't think I linked the new document to anything... I did link _from_ the new document to existing ones.

jorgeorpinel commented:
The tests can be accessed from the top tab of the PR:

![image](https://user-images.githubusercontent.com/1477535/91881775-a5fee500-ec47-11ea-9998-ff7b199c0123.png)

Or from it's bottom section:

![image](https://user-images.githubusercontent.com/1477535/91881704-84056280-ec47-11ea-99c6-250eb2ade2ef.png)

> Could it be related to changes in:

Nah, that would fail locally. Like I mentioned it's related to a link-checking test that only exists on the CI, which is a known bug of that script so nothing to worry about for now.

jorgeorpinel commented:
OK I deployed your use case (see https://dvc-landing-feat-ci-use-phxvgz.herokuapp.com/doc/use-cases/ci-and-ml) and will make a more comprehensive review now @drorata. Thanks for all the updates so far.

BTW please notice the formatting is currently broken (you can see the `yarn-format` test failing, as well as the Restyled integration activating automatically).

drorata commented on content/docs/use-cases/ci-and-ml.md:
```diff
@@ -0,0 +1,172 @@
+ML engineers and data scientists face many unique challenges. Some of these
+challenges can be solved by adopting well-established concepts from traditional
+software engineering and development. Probably the most common adoption is
+versioning control. Another adoption, unfortunately less common though, could be
+unit testing the various stages of a data processing pipeline. Here is a non
+exhaustive list of a few other unique challenges:
+
+- automatically re-train models upon changes
+- systematically compare (and report) performance of different versions of a
+  model
+- test data correctness when new datasets are introduced
+- handle a model training that is computationally intensive
```
No capitalization nor punctuation. @jorgeorpinel anything else here?

jorgeorpinel replied:
Capital letters are optional indeed according to our guidelines but we don't really use this format anywhere. No worries, I'll update it for you bellow so we can move on:

```suggestion
versioning control. Another adoption, unfortunately less common though, could be
unit testing the various stages of a data processing pipeline. Here is a non
exhaustive list of a few other unique challenges:

- Automatically re-train models upon changes
- Systematically compare (and report) performance of different versions of a
  model
- Test data correctness when new datasets are introduced
- Handle a model training that is computationally intensive
```

drorata commented on content/docs/use-cases/ci-and-ml.md:
```diff
@@ -0,0 +1,172 @@
+tackling the aforementioned problems and many others. In this article, you will
+learn about a few CI use cases utilizing `dvc` and `cml`.
+
+_Remarks:_ The code snippets in this article are for demonstration purposes and
+not actually running snippets. Furthermore, it is assumed that the `git`
+repository is hosted on `GitHub`; the same ideas can be implemented when using
+`GitLab` or any other hosting solution that suports CI.
+
```
I think it is important to keep the remark at the beginning and align the expectations with the reader. I made some changes to the remark and turned it into a quote block. @jorgeorpinel feel free to resolve if nothing else is needed.

jorgeorpinel replied:
I see they're gone, thanks. I think the note could be shorter still, but OK resolving this.

drorata commented on content/docs/use-cases/ci-and-ml.md:
```diff
@@ -0,0 +1,172 @@
+- Test data correctness when new datasets are introduced.
+- Handle a model training that is computationally intensive.
+
+Combining a version control tool (like `git`) together with `dvc` and `cml`
+opens a wide spectrum of possibilities that can help building better and smarter
+data products. Adding [CI](https://en.wikipedia.org/wiki/Continuous_integration)
+concepts to this triple allows the bulding and **coding** of flows that can help
+tackling the aforementioned problems and many others. In this article, you will
+learn about a few CI use cases utilizing `dvc` and `cml`.
```
I removed the last sentence. I would like to keep the existing **Bonuses**. What about the "What's next"? I give a hint to another use case there and I think it's worthy to keep. I will revise it and your feedback would be much appreciated.

drorata replied:
> I found this paragraph too indirect. I could be brief and to the point, which AFAIK is: 1. We propose CI as a solution for the challenges and 2. This can be done combining Git, DVC, and CML, as will be shown ahead.

In my opinion, this paragraph is already rather short and fits nicely with the introduction. I wouldn't like to replace it with bullet points.

jorgeorpinel replied:
OK, no strong opinion about Bonus sections ⚡ just very slightly concerned about consistency in structure among our use cases.

I see the paragraph has been completely rewritten too, so resolving this.

drorata commented on content/docs/use-cases/ci-and-ml.md:
```diff
@@ -0,0 +1,172 @@
+ML engineers and data scientists face many unique challenges. Some of these
+challenges can be solved by adopting well-established concepts from traditional
+software engineering and development. Probably the most common adoption is
+versioning control. Another adoption, unfortunately less common though, could be unit testing
```
I disagree that it is distracting. You can still meet data scientists who do not version control their work let alone write proper unit tests for their code.

@jorgeorpinel what do you think about the changes? Can we resolve this one?

jorgeorpinel replied:
I see this part has evolved ans is being reviewed by Elle now 👍

> For the record what I meant by distracting is that Use Cases are meant to be as simple as possible in terms of the explanations. It's tricky but important to leave out many of the details. Trust me, this is hard, I struggle with it constantly... Resolving.

drorata commented on content/docs/use-cases/ci-and-ml.md:
```diff
@@ -9,11 +9,11 @@ versioning control. Another adoption, unfortunately less common though, could be
-- test data correctness when new datasets are introduced
-- handle a model training that is computationally intensive
+- Test data correctness when new datasets are introduced
+- Handle a model training that is computationally intensive
```
@jorgeorpinel I understand it that it should either be:

* capitalization _and_ punctuation, or
* no capitalization and no punctuation.

Your change is a mix. Isn't it?

jorgeorpinel replied:
It's fine to capitalize without an ending period. Optional but it's what we use so far everywhere so recommended, for consistency.

drorata commented on content/docs/use-cases/ci-and-ml.md:
```diff
@@ -0,0 +1,169 @@
+- Systematically compare (and report) performance of different versions of a
+  model
+- Test data correctness when new datasets are introduced
+- Train a model which is computationally intensive or requires special
+  environment (e.g. GPU)
```
@jorgeorpinel is it clearer now?

drorata commented:
@jorgeorpinel see inline:

> OK my comprehensive feedback:
> 
> * On the highest level I think the structure is great and makes sense 👍

Thanks!

> * The intro is good but could be shorter while keeping the same key info. I'm leaving a detailed review with multiple comments about it actually (listed after this long comment) — probably too low level in this case 😅 sorry...

I addressed your comments and I hope it is more to your liking now. I think it's good to have in this case a somewhat high-level intro because the topic is not to some extent a surprising(?) twist in the plot around DVC. 

> * It's not super clear which of the challenges form the intro do the 2 technical sections map to — not sure this is a big deal, depends on what your intentions are.

Again, I feel it makes sense to have it this way so we start with a broader scope and give two examples derived from this scope.

> * In general I see many opportunities to trim sentences and paragraphs to be more direct, less redundant, and sometimes easier to read, even if it means having to omit details and side notes.

Can you be more concrete?

>   > More links from specialized terms and DVC concepts could replace details too.
> * On the other hand, DVC commands and workflow details are somewhat lacking. Maybe we're relying too much on textual explanation and too little on demonstrative blocks? (More comments on this below).

This is a high-level use case and thus it cannot be too technical. As per the discussion I had with @shcheklein, there is no proper working example backing up this discussion. It is intended to outline the possibilities and not to provide cookbook recipes.

> * **Automatic model training**:
>   
>   * The explanations combine several complex ideas and jargon which can be confusing e.g. model training, "test suite", CI, "remote cache", etc. I'd try to split into shorter paragraphs (one per point) and leave out some of the concepts, if not central to getting the main point across.

I have in mind a mixed audience here; data scientists as well as maybe DevOps folks. I read it again and I think that both target groups (and others as well) can easily relate to the concepts. Moreover, I like to stretch my knowledge when reading use cases, so I can be pointed (even implicitly) to further learning.

>   * DVC usage, even when hypothetical, should be a little clearer, I think. What comes first and why, run? pull? repro? In which machine? Try to explain the workflow and command effects if possible. Feel free to use short console code blocks (see existing use cases).

I don't. Can you be more specific? This use case, to some degree, assumes some knowledge of DVC, so I'm not sure it makes sense to add this layer of details.

> * **Seamless model evaluation**: similarly, what's the DVC workflow? How/when does CML interact with it? Any relevant links to include ([e.g.](https://github.com/iterative/cml#using-cml-with-dvc))? Let's try to help the reader connect the dots.

That's a good point. The snippets in the link you mention inspire this article. I will add the link you suggested. Otherwise, I think this would be a greater effort, beyond the scope of this PR to streamline the documentation of DVC and CML.

> * Finally, are those 2 technical sections comprehensive? I.e. do they cover 90% of the scenarios we expect for people interested in DVC+CI? Please decide with Ivan or/and Elle.

@shcheklein @elleobrien ? What do you think?

elleobrien commented:
Taking a look now! Here are my thoughts-

1. I don't believe CML has ever been mentioned in the DVC docs until now, so it's not a given that readers will even be aware of it. The introduction reads as if CML, DVC, and CI are a trio- but CML is a toolkit for doing CI, not an alternative. I think we need to refine these ideas slightly.  Maybe something like:

```One of the core principles of DevOps is continuous integration (CI), which briefly means: frequent code, data, and model check-ins, automatic testing, and fast feedback. DVC provides a solid basis for automating model training and testing with constructs like pipelines and metrics. To implement a fully functioning CI system, which responds to changes in your project with automatically run tests and feedback, we've built a complementary tool, Continuous Machine Learning (CML). CML helps you use DVC with GitHub Actions & GitLab CI.```

Or something like that. And then go into the use cases.

2. I think we need a sentence in the intro explaining the relationship of the blog sections to the big idea. Something like, `Two of the most common scenarios for using CI are automatic model training and model testing.` Right now, it's not clear how they logically follow from the intro.

3. The "Bonus" sections are not a bad idea, but I suspect we're missing the ideal formatting element. Using a header looks a little weird, since they're too short to be standalone paragraphs. Ideally, we'd have a little colored box, kind of like in textbooks :) Since we don't have that currently, I'm wondering what formatting we could use so it doesn't look odd. Maybe **Bonus** with a lightning bolt emoji? 

4. There's still a missing element about why model retraining is hard. From the user's standpoint, it might only mean running a script. So I maybe we want to make a point that we want to automate model retraining for the sake of reducing human error/having a complete log of project. 

These are my thoughts so far!

drorata commented:
Thanks, @elleobrien 👍 See inline:

> Taking a look now! Here are my thoughts-
> 
> 1. I don't believe CML has ever been mentioned in the DVC docs until now, so it's not a given that readers will even be aware of it. The introduction reads as if CML, DVC, and CI are a trio- but CML is a toolkit for doing CI, not an alternative. I think we need to refine these ideas slightly.  Maybe something like:
> 
> `One of the core principles of DevOps is continuous integration (CI), which briefly means: frequent code, data, and model check-ins, automatic testing, and fast feedback. DVC provides a solid basis for automating model training and testing with constructs like pipelines and metrics. To implement a fully functioning CI system, which responds to changes in your project with automatically run tests and feedback, we've built a complementary tool, Continuous Machine Learning (CML). CML helps you use DVC with GitHub Actions & GitLab CI.`

I like this addition. Do I understand correctly and you'd suggest to add it here (I don't know why doesn't it show the snippet in this comment):

https://github.com/drorata/dvc.org/blob/5436941e80b2a9adf2936ab98f450ebf83d0ca0e/content/docs/use-cases/ci-and-ml.md#L18

> 
> Or something like that. And then go into the use cases.
> 
> 1. I think we need a sentence in the intro explaining the relationship of the blog sections to the big idea. Something like, `Two of the most common scenarios for using CI are automatic model training and model testing.` Right now, it's not clear how they logically follow from the intro.

I don't understand why... The first two examples in the list present in the intro are tightly coupled to the two use cases discussed later. Isn't it?

> 2. The "Bonus" sections are not a bad idea, but I suspect we're missing the ideal formatting element. Using a header looks a little weird, since they're too short to be standalone paragraphs. Ideally, we'd have a little colored box, kind of like in textbooks :) Since we don't have that currently, I'm wondering what formatting we could use so it doesn't look odd. Maybe **Bonus** with a lightning bolt emoji?

I'll push this. What do you think?

> 3. There's still a missing element about why model retraining is hard. From the user's standpoint, it might only mean running a script. So I maybe we want to make a point that we want to automate model retraining for the sake of reducing human error/having a complete log of project.

When balancing between conciseness and detailing, I think the bullet point about this point in the intro touch bases two central concerns (complexity and environment). The use case later caters to these two issues. Moreover, I think we can assume that the target audience is familiar with these difficulties.

elleobrien commented:
@drorata here are my thoughts:

- On explaining what the two sections (model training and testing) have to do with the logic of the whole article, I think it's necessary even though they're mentioned in the bulleted list. The bulleted list establishes problems common to ML. But the article is about ways DVC and CML can address these problems. I think a simple sentence about your intent would help.

- Even if it's not strictly necessary to rehash what about training models is hard, I think the sentence 

> Training a model is often a very demanding process. 
is not very interesting. Even a quick detail about why it's demanding would make the writing stronger. 

Generally, I think of our "target audience" as broad. We are often speaking to data scientists, engineers, and project managers (of varying technical levels), and in the ML tool space, telling tools apart is difficult- "what makes DVC different from MLFlow or KubeFlow or ____ ?" is the question I get asked more than anything. That's why I tend to err on the side of specificity and even repetition. Of course, that's only my preference :)

drorata commented:
@elleobrien I introduced two more changes (the paragraph right after the bullet points and the beginning of the first example). Let me know what do you think.

pared commented on content/docs/use-cases/ci-and-ml.md:
```diff
@@ -0,0 +1,172 @@
+
+jobs:
+  train:
+    runs-on: ubuntu-latest
```
I presume that this is showcase for user, so that one can reproduce it by her/him self. In that case I would recommend sticking to particular version of system, so that we eliminate possibility of some dependencies/ environmental setup being different among images.

drorata replied:
@pared do you mean the version of Ubuntu? As the disclaimer states at the beginning, the snippets are for demonstration purposes - I expect readers to be inspired by this and work out the details as per their need. Therefore, although I totally understand that versions have to be included in the environment's definitions, I don't think it is important here. 

See my other comment. I'll mention you there :)

pared replied:
Yeah, I guess I did look at it too much from our tests/benchmarks perspective - where we don't want any unnecessary changes. In this case that makes perfect sense to use the latest version.

pared commented on content/docs/use-cases/ci-and-ml.md:
```diff
@@ -0,0 +1,172 @@
+
+jobs:
+  evaluate:
+    runs-on: ubuntu-latest
```
Same as above.

drorata replied:
Let's resolve it together with the other one

pared reviewed (approved):
Looks very good and intuitive to me. One comment regarding reproducibility.

elleobrien commented on content/docs/use-cases/ci-and-ml.md:
````diff
@@ -0,0 +1,172 @@
+          dvc push
+```
+
+**Bonus (⚡️):** It might be that the training needs a tailored environment.
````
How about **Bonus ⚡️**? remove parentheses

drorata replied:
Better?

elleobrien replied:
Yep!

elleobrien commented on content/docs/use-cases/ci-and-ml.md:
```diff
@@ -0,0 +1,172 @@
+DVC into CI tools we've built a complementary tool, Continuous Machine Learning
+([CML](http://www.cml.dev)). Putting these tools together allows the bulding and
+**coding** of flows that can help tackling the aforementioned problems and many
+others. We will next dive a little deeper and outline how to harness these tools
```
Looks good! Let's add a line break somewhere for readability- maybe put "We will next dive a little deeper...." after a break?

drorata replied:
This will become a paragraph with a single sentence. Is it desirable?

elleobrien replied:
Personally I am okay with it! I use single sentence paragraphs. It is a matter of personal preference.

elleobrien commented on content/docs/use-cases/ci-and-ml.md:
```diff
@@ -0,0 +1,172 @@
+# CI and ML
+
+ML engineers and data scientists face many unique challenges. Some of these can
```
We can make the first sentence a little more interesting- maybe something along the lines of, "ML engineers and data scientists face many challenges owing to the complexity of the typical project." What do you think?

drorata replied:
I agree that this sentence can be rephrased in a more appealing way. But the challenges are not due to the "complexity". After all, non-data/ML software projects can be very complex as well.

Maybe: "ML engineers and data scientists face many challenges derived from the dependency on data and complexity of models and other reasons"?

elleobrien replied:
That would be better- 

FYI, I mean complexity not as in cyclomatic or algorithmic complexity, but as in system complexity (see http://papers.nips.cc/paper/5656-hidden-technical-debt-in-machine-learning-systems.pdf). Technical debt due to hardware, data dependencies, feedback loops, etc.

shcheklein commented:
Closing this for now as we discussed with @drorata it might take another iteration to refactor it.

jorgeorpinel commented:
Hi! Some very late last replies from me:

> > DVC commands and workflow details are somewhat lacking. Maybe we're relying too much on textual explanation and too little on demonstrative blocks?
> 
> This is a high-level use case and thus it cannot be too technical... It is intended to outline the possibilities and not to provide cookbook recipes.

Right. I did not say we want recipes. I'm only suggesting demo code blocks can be better for the technical parts that are hard and complicated to explain in text. We do want to show how DVC commands work for this use case though, IMO.

> > The explanations combine several complex ideas and jargon which can be confusing e.g. model training, "test suite", CI, "remote cache", etc. I'd try to split into shorter paragraphs...
>
> I have in mind a mixed audience here; data scientists as well as maybe DevOps folks... I think that both target groups (and others as well) can easily relate to the concepts. Moreover, I like to stretch my knowledge when reading use cases

You may be assuming a bit much about the reader. It's safer to keep the language simple and direct when possible, I think. In any case, my feedback was more on the writing itself: Maybe all these concepts have to be presented, but most likely not in the same sentence, often not even paragraph.

> This use case, to some degree, assumes some knowledge of DVC

I wouldn't say use cases assume knowledge of DVC, other than what's shown in https://dvc.org/, https://dvc.org/doc, and the https://dvc.org/doc/use-cases index page itself. But not really about its usage at all really.

Thanks, looking forward to the next iteration.
